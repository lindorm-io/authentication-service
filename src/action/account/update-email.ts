import Joi from "@hapi/joi";
import { IAuthContext } from "../../typing";
import { JOI_EMAIL } from "../../constant";

export interface IUpdateAccountEmailOptions {
  updatedEmail: string;
}

const schema = Joi.object({
  updatedEmail: JOI_EMAIL,
});

export const updateAccountEmail = (ctx: IAuthContext) => async (options: IUpdateAccountEmailOptions): Promise<void> => {
  await schema.validateAsync(options);

  const { account, logger, repository } = ctx;
  const { updatedEmail } = options;

  account.email = updatedEmail;

  await repository.account.update(account);

  logger.debug("device email updated", {
    accountId: account.id,
    email: updatedEmail,
  });
};
