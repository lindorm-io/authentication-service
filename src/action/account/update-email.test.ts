import MockDate from "mockdate";
import { updateAccountEmail } from "./update-email";
import { getTestRepository, getTestAccount, inMemoryStore, logger, resetStore } from "../../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("../../support", () => ({
  assertBearerTokenScope: jest.fn(() => () => {}),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("updateAccountEmail", () => {
  let ctx: any;

  beforeEach(async () => {
    ctx = {
      account: getTestAccount("email@lindorm.io"),
      logger,
      repository: await getTestRepository(),
    };
    await ctx.repository.account.create(ctx.account);
  });

  afterEach(resetStore);

  test("should update account email", async () => {
    await expect(
      updateAccountEmail(ctx)({
        updatedEmail: "email@lindorm.io",
      }),
    ).resolves.toBe(undefined);
    expect(inMemoryStore).toMatchSnapshot();
  });
});
