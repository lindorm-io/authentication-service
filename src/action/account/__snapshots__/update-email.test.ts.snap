// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateAccountEmail should update account email 1`] = `
Object {
  "authentication": Object {
    "account": Object {
      "data": Array [
        Object {
          "created": 2020-01-01T07:00:00.000Z,
          "email": "email@lindorm.io",
          "events": Array [
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "account_created",
              "payload": Object {
                "created": 2020-01-01T07:00:00.000Z,
                "email": "email@lindorm.io",
                "otp": Object {
                  "signature": null,
                  "uri": null,
                },
                "password": Object {
                  "signature": null,
                  "updated": null,
                },
                "permission": "user",
                "updated": 2020-01-01T07:00:00.000Z,
              },
            },
            Object {
              "date": 2020-01-01T07:00:00.000Z,
              "name": "account_email_changed",
              "payload": Object {
                "email": "email@lindorm.io",
              },
            },
          ],
          "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
          "identityLinked": false,
          "otp": Object {
            "signature": null,
            "uri": null,
          },
          "password": Object {
            "signature": null,
            "updated": null,
          },
          "permission": "user",
          "updated": 2020-01-01T07:00:00.000Z,
          "version": 1,
        },
      ],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;
